crates/generate/src/funcs.rs

```diff
@@ -242,6 +242,42 @@ fn marshal_arg(
                 };
             }
         }
+        witx::Type::Array(arr) => {
+            let lifetime = anon_lifetime();
+            let pointee_type = names.type_ref(arr, lifetime.clone());
+            let ptr_name = names.func_ptr_binding(&param.name);
+            let len_name = names.func_len_binding(&param.name);
+            let name = names.func_param(&param.name);
+            quote! {
+                let num_elems = match memory.ptr::<u32>(#len_name as u32) {
+                    Ok(p) => match p.as_ref() {
+                        Ok(r) => r,
+                        Err(e) => {
+                            #error_handling
+                        }
+                    }
+                    Err(e) => {
+                        #error_handling
+                    }
+                };
+                let #name = match memory.ptr_mut::<wiggle_runtime::GuestPtrMut<#lifetime, #pointee_type>>(#ptr_name as u32) {
+                    Ok(p) => match p.read_ptr_from_guest() {
+                        Ok(r) => match r.array_mut(*num_elems) {
```

Comment:
I think this is incorrect - #ptr_name is itself a pointer, but this uses it as a pointer to a pointer. I think this should be `memory.ptr_mut::<#pointee_type>(#ptr_name)`. The `read_ptr_from_guest` clause can go away completely.

The other question I have is that this is creating an ArrayMut for an argument-position array. All of the arrays we define in wasi for args are immutable. I think we should treat all witx `(array <pointee>)` as immutable arrays. Mutable arrays are still useful for processing the contents of an iovec struct - we are able to construct them in the wiggle user's code from a GuestPtrMut (which in turn comes from an `(@witx pointer <pointee>)`) and a length.

Reply:
Hmm, so let me explain my reasoning about ptr-to-ptr. From what I understood and how we handle `PointerLengthPair`, the layout for an array in guest would be something like:

```
| 0x8 | 0x0 | 0x0 | 0x0 | 0x4 | 0x0 | 0x0 | 0x0 | 0x1 | 0x1 | 0x1 | 0x1 | ...
```

where the first 4 bytes are a pointer to the address of the array (in this case, we point at `0x8` as the start of the array), the next four bytes are the length (represented as `u32`), and the next four bytes are the actual elements of the array. Otherwise, if we had one less level of indirection, then the pointer *would* be the start of the array. This might be the way it's supposed to be, but then how is the layout handled in memory? Does that mean that guest writes into memory first all of the elements and then writes the length of the array at some different address?

Reply:
As far as treating all arrays as immutable, that's fine, but then how do we convert an immutable array to a mutable array for `iovec`? Because the `wiggle-generate` crate will type both `iovec` and `ciovec` arrays as either immutable or mutable, right? Or am I missing something here?

Reply:
To put into more immediate perspective, note the difference between `populate_excuses` and `reduce_excuses` interface funcs. It is true that I could have spec out a different type for a *mutable* array as an input to `populate_excuses`, but even then, I cannot tell programmatically which is one is which at the `wiggle-generate` level.

Reply:
Thanks for explaining, I now see exactly where our thoughts on this diverged. Witx is strange in that we pass everything that doesn't fit in a single ABI argument by reference, except for arrays, which we pass as a pair of arguments:
 https://github.com/cranestation/wasi-libc/blob/5785f39fbeda720a3d07ddc3bdc8979b390faed6/libc-bottom-half/headers/public/wasi/api.h#L1926

I think what happened here is that the C ABI originally passed a pointer/len pair of arguments here, and then ret-conned those into a single array-typed parameter, but defined ABI as two arguments (instead of one pointer to the pointer/length pair) because witx was supposed to be descriptive of snapshot 0, rather than prescriptive.

So far in WASI, we only ever pass arrays as arguments, we never use them inside any other other data structure (behind a pointer e.g. (@witx pointer (array ...)), or inside a struct or anything like that). We'll lay them out in memory like you describe if we ever do put them inside something else.

As far as `iovec_array` and `ciovec_array` - the array itself is immutable, meaning that the memory of its elements is not modified during the call. The elements themselves contain either a `(@witx pointer u8)` or `(@witx const_pointer u8)`, which determines whether *that* pointee is mutable or not. Hence, we could swap out the entire `ciovec` definition from a `(struct (field $buf (@witx const_pointer u8)) (field $buv_len $size))` to an `(array u8)` in witx, and be ABI compatible. (Just like in the tagged union transformation I have underway, we'd also teach all of the tools in the ecosystem that the memory layout of `array` works exactly like that struct did, so they remained compatible when they slurp in the new witx files.) However, with the current mutability convention, we can't swap out the definiton of `iovec` in the same way. Maybe we can define a new type `(@witx mutable_array <pointee>)` that works just like iovec (whose pointee is u8). That is useful elsewhere too - we could then rewrite arguments like the `buf`/`buf_len` pair in `path_readlink` to be a mutable array as well.

Reply:
Right, gotcha, everything makes perfect sense now, thanks! Let me commit the changes which conform to how it should be, and I'll re-ping you for a review.